src/thor/timedistancematrix.cc

```diff
@@ -141,416 +203,217 @@ void TimeDistanceMatrix::ExpandForward(GraphReader& graphreader,
-    // Copy the EdgeLabel for use in costing
-    EdgeLabel pred = edgelabels_[predindex];
+      // Terminate when we are beyond the cost threshold
+      if (pred.cost().cost > current_cost_threshold_) {
+        one_to_many = FormTimeDistanceMatrix();
+        break;
+      }
 
-    // Mark the edge as permanently labeled. Do not do this for an origin
-    // edge (this will allow loops/around the block cases)
-    if (!pred.origin()) {
-      edgestatus_.Update(pred.edgeid(), EdgeSet::kPermanent);
+      // Expand forward from the end node of the predecessor edge.
+      Expand(graphreader, pred.endnode(), pred, predindex, false);
     }
 
-    // Identify any destinations on this edge
-    auto destedge = dest_edges_.find(pred.edgeid());
-    if (destedge != dest_edges_.end()) {
-      // Update any destinations along this edge. Return if all destinations
-      // have been settled.
-      tile = graphreader.GetGraphTile(pred.edgeid());
-      const DirectedEdge* edge = tile->directededge(pred.edgeid());
-      if (UpdateDestinations(dest, locations, destedge->second, edge, tile, pred, matrix_locations)) {
-        return FormTimeDistanceMatrix();
+    // Insert one-to-many into many-to-many
+    if (FORWARD) {
+      for (size_t target_index = 0; target_index < destinations.size(); target_index++) {
+        size_t index = origin_index * origins.size() + target_index;
+        many_to_many[index] = one_to_many[target_index];
+      }
+    } else {
+      for (size_t source_index = 0; source_index < destinations.size(); source_index++) {
+        size_t index = source_index * origins.size() + origin_index;
+        many_to_many[index] = one_to_many[source_index];
       }
     }
-
-    // Terminate when we are beyond the cost threshold
-    if (pred.cost().cost > current_cost_threshold_) {
-      return FormTimeDistanceMatrix();
-    }
-
-    // Expand forward from the end node of the predecessor edge.
-    ExpandReverse(graphreader, pred.endnode(), pred, predindex, false);
+    clear();
   }
-  return {}; // Should never get here
+
+  return many_to_many;
 }
 
-// Many to one time and distance cost matrix. Computes time and distance
-// from many locations to many locations.
-std::vector<TimeDistance> TimeDistanceMatrix::ManyToMany(
-    const google::protobuf::RepeatedPtrField<valhalla::Location>& locations,
-    GraphReader& graphreader,
+template std::vector<TimeDistance> TimeDistanceMatrix<ExpansionType::forward>::SourceToTarget(
+    const google::protobuf::RepeatedPtrField<valhalla::Location>& source_location_list,
+    const google::protobuf::RepeatedPtrField<valhalla::Location>& target_location_list,
+    baldr::GraphReader& graphreader,
     const sif::mode_costing_t& mode_costing,
     const sif::travel_mode_t mode,
-    const float max_matrix_distance) {
-  return SourceToTarget(locations, locations, graphreader, mode_costing, mode, max_matrix_distance);
-}
```

Comment:
Also removed the `ManyToMany`, it wasn't ever used and seemed pretty pointless (though it was public API.. as I understand we don't necessarily care about the C++ API while versioning?)